File: Double a Number Represented as a Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func doubleIt(head *ListNode) *ListNode {
    dummyHead := &ListNode{Next:head}
    prev := dummyHead
    ptr := head
    for ptr != nil {
        ptr.Val *= 2
        prev.Val += ptr.Val / 10
        ptr.Val %= 10
        prev = ptr
        ptr = ptr.Next
    }

    if dummyHead.Val != 0 {
        return dummyHead
    }
    return dummyHead.Next
}

File: Double a Number Represented as a Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* doubleIt(ListNode* head) {
        ListNode* dummyHead = new ListNode(0, head);
        ListNode* prev = dummyHead;
        ListNode* ptr = head;
        while (ptr != nullptr) {
            ptr->val *= 2;
            prev->val += ptr->val / 10;
            ptr->val %= 10;
            prev = ptr;
            ptr = ptr->next;
        }
        if (dummyHead->val != 0) {
            return dummyHead;
        }
        return dummyHead->next;
    }
};